src/lxcfs.c

```diff
@@ -771,6 +771,13 @@ static int lxcfs_releasedir(const char *path, struct fuse_file_info *fi)
 {
 	int ret;
 
+        if (path == NULL) {
+                up_users();
+                ret = do_cg_releasedir(path, fi);
```

Comment:
why we are calling `do_cg_releasedir` here? Why not `do_sys_releasedir`? Without `path` we can't determine what to call here. I think we have to print out some error to the LXCFS log and return `-EINVAL`.

Reply:
>I am (potentially unsafely) assuming that whatever user data exists in fi will always be the same struct being released here.

Yep, that's what I'm trying to say. Right now that makes no difference to call `do_cg_releasedir` or `do_sys_releasedir` but in the future, this may be changed. We can't rely on that.

>The problem with not doing anything at all is may end up just slowly leaking memory.

Yep, but a memory leak combined with the warning message in logs is better than potentially incorrect memory free ;-)

>Passing -EINVAL will result in unexpected behaviour in userspace applications.

I've checked how this retvalue is handled in libfuse... and... it's just ignored. :-)
https://github.com/libfuse/libfuse/blob/33736958b61d90d9db9b03441103035316f09abc/lib/fuse.c#L1982
Take a look on `fuse_fs_releasedir()` calls.

>Spitting a log output might keep the process running but its pretty unlikely with the information kept in the userspace structure will reveal what the original path was that was rmdir'd. You could extend the userspace structure to add tracking though.

Yep, but this warning will be a sign of a problem. And yes, extending the log message with some additional info, in this case, is a good idea.

Reply:
> I've checked how this retvalue is handled in libfuse... and... it's just ignored. :-)
Thats great! I agree in that case leak/EINVAL is nicer.

If I can get the time approved I shall try to work out what the condition is that results in this behaviour as it might reveal something more fundamental going on.

Reply:
BTW, we can use `(struct file_info)->type` to detect what to call `do_cg_releasedir` or `do_sys_releasedir`... Then we don't need to know the `path` at all in the release* callbacks!

Reply:
With a bit of hacking, I managed to resolve the userspace component of the coredumps I have in the backtraces.
These are 3 different crash examples.

```
(gdb) print *(struct file_info *)fi->fh                                                                 
$2 = {controller = 0x7f0408105af0 "systemd", 
  cgroup = 0x7f04081075e0 "lxc.payload.infra-app-test-31/system.slice/snap.mount", file = 0x0 <foo>, 
  type = 0, buf = 0x0 <foo>, buflen = 0, size = 779116910, cached = 1853189997}
```

And

```
(gdb) print *(struct file_info *)fi->fh
$1 = {controller = 0x7f02f4106a50 "devices", cgroup = 0x7f02f4118740 "lxc.payload.infra-app-test-33/system.slice/sssd-pac.socket", file = 0x0 <foo>, type = 0, buf = 0x0 <foo>, buflen = 0, 
  size = 778989426, cached = 1987208563}
```

And

```
(gdb) print *(struct file_info *)fi->fh
$4 = {controller = 0x7f9f90000ca0 "systemd", cgroup = 0x7f9f9000d470 "lxc.payload.infra-app-test-31/system.slice/system-getty.slice", file = 0x0 <foo>, type = 0, buf = 0x0 <foo>, buflen = 0, 
  size = 779318626, cached = 1987208563}
```

We use a lot of cgroup v1 namespaces on these hosts, however (which I have no backtraces for unfortunately) we have another that exhibits the same behaviour that is using cgroupv2 only).

In any case they are always cgroup files that crash in my case.

We pretty much can hammer our hosts (having 150+ active containers at at time).

I will update the PR with code tomorrow that handles all 3 cases cleanly.